packages/pigeon/lib/kotlin_generator.dart

```diff
@@ -66,309 +66,585 @@ class KotlinOptions {
+          }
+        });
       });
-    }
-  });
+    });
+  }
+
+  /// Writes the codec class that will be used by [api].
+  /// Example:
+  /// private static class FooCodec extends StandardMessageCodec {...}
+  void _writeCodec(Indent indent, Api api, Root root) {
+    assert(getCodecClasses(api, root).isNotEmpty);
+    final Iterable<EnumeratedClass> codecClasses = getCodecClasses(api, root);
+    final String codecName = _getCodecName(api);
+    indent.writeln('@Suppress("UNCHECKED_CAST")');
+    indent.write('private object $codecName : StandardMessageCodec() ');
+    indent.scoped('{', '}', () {
+      indent.write(
+          'override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? ');
+      indent.scoped('{', '}', () {
+        indent.write('return when (type) ');
+        indent.scoped('{', '}', () {
+          for (final EnumeratedClass customClass in codecClasses) {
+            indent.write('${customClass.enumeration}.toByte() -> ');
+            indent.scoped('{', '}', () {
+              indent.write('return (readValue(buffer) as? List<Any?>)?.let ');
+              indent.scoped('{', '}', () {
+                indent.writeln('${customClass.name}.fromList(it)');
+              });
+            });
+          }
+          indent.writeln('else -> super.readValueOfType(type, buffer)');
+        });
+      });
+
+      indent.write(
+          'override fun writeValue(stream: ByteArrayOutputStream, value: Any?) ');
+      indent.writeScoped('{', '}', () {
+        indent.write('when (value) ');
+        indent.scoped('{', '}', () {
+          for (final EnumeratedClass customClass in codecClasses) {
+            indent.write('is ${customClass.name} -> ');
+            indent.scoped('{', '}', () {
+              indent.writeln('stream.write(${customClass.enumeration})');
+              indent.writeln('writeValue(stream, value.toList())');
+            });
+          }
+          indent.writeln('else -> super.writeValue(stream, value)');
+        });
+      });
+    });
+    indent.addln('');
+  }
+
+  void _writeWrapResult(Indent indent) {
+    indent.addln('');
+    indent.write('private fun wrapResult(result: Any?): List<Any?> ');
+    indent.scoped('{', '}', () {
+      indent.writeln('return listOf(result)');
+    });
+  }
+
+  void _writeWrapError(Indent indent) {
+    indent.addln('');
+    indent.write('private fun wrapError(exception: Throwable): List<Any> ');
+    indent.scoped('{', '}', () {
+      indent.write('return ');
+      indent.scoped('listOf<Any>(', ')', () {
+        indent.writeln('exception.javaClass.simpleName,');
+        indent.writeln('exception.toString(),');
+        indent.writeln(
+            '"Cause: " + exception.cause + ", Stacktrace: " + Log.getStackTraceString(exception)');
+      });
+    });
+  }
+
+  @override
+  void finalWriteFile(KotlinOptions generatorOptions, Root root,
+      StringSink sink, Indent indent) {
+    _writeWrapResult(indent);
+    _writeWrapError(indent);
```

Comment:
I think a `writeGeneralUtilities` (probably near the start of the file) or something where we can put miscellaneous general-purpose classes and/or functions might be useful, rather than them having to be in a random place, could be useful.

Reply:
changed